File: examples/durability/retry-failure/python/agent.py
"""Webhook Receiver Agent — durability/retry-failure example.

Delivery: stream  (GET /v1/agents/{address}/intents/stream)

Processes incoming webhook events. In this durability test scenario:
- Run WITH the agent: intent is processed normally → COMPLETED
- Run WITHOUT the agent: AXME retries delivery up to max_delivery_attempts,
  then transitions to FAILED with `intent.delivery_failed` event

Run (normal completion):
    export AXME_API_KEY=<your-api-key>
    export AXME_AGENT_ADDRESS=agent://<org>/<workspace>/webhook-receiver-agent
    python agent.py
    axme scenarios apply examples/durability/retry-failure/scenario.json --watch

Run (to observe retry behavior):
    # Do NOT start agent.py
    axme scenarios apply examples/durability/retry-failure/scenario.json --watch
    # Watch AXME retry delivery and eventually emit intent.delivery_failed
"""
from __future__ import annotations

import hashlib
import logging
import os
import sys
from typing import Any

from axme import AxmeClient, AxmeClientConfig

logging.basicConfig(
    level=logging.INFO,
    format="%(asctime)s  [%(levelname)s]  %(message)s",
    datefmt="%H:%M:%S",
)
log = logging.getLogger("webhook-receiver")

AXME_BASE_URL      = os.environ.get("AXME_BASE_URL", "https://api.cloud.axme.ai")
AXME_API_KEY       = os.environ.get("AXME_API_KEY", "")
AXME_AGENT_ADDRESS = os.environ.get("AXME_AGENT_ADDRESS", "webhook-receiver-agent")

if not AXME_API_KEY:
    sys.exit("AXME_API_KEY is required")


def _process_webhook(payload: dict[str, Any]) -> dict[str, Any]:
    """Process incoming webhook event — real idempotent processing."""
    event_type = payload.get("event_type", "unknown")
    source     = payload.get("source", "")
    test_id    = payload.get("test_id", "")

    # Generate a deterministic fingerprint for idempotency tracking
    fingerprint = hashlib.sha256(
        f"{event_type}:{source}:{test_id}".encode()
    ).hexdigest()[:16]

    return {
        "event_type":   event_type,
        "source":       source,
        "fingerprint":  fingerprint,
        "processed":    True,
        "note":         f"webhook event {event_type!r} processed from {source!r}",
    }


def handle_intent(client: AxmeClient, delivery: dict[str, Any]) -> None:
    intent_id = delivery.get("intent_id", "")
    if not intent_id:
        log.warning("received delivery without intent_id, skipping")
        return

    log.info("received intent %s", intent_id)

    try:
        resp   = client.get_intent(intent_id)
        intent = resp.get("intent") or resp
    except Exception as exc:
        log.error("get_intent(%s) failed: %s", intent_id, exc)
        return

    status = (intent.get("lifecycle_status") or intent.get("status") or "").upper()
    if status not in {"CREATED", "DELIVERED", "ACKNOWLEDGED", "IN_PROGRESS", "WAITING"}:
        log.debug("skipping intent %s with status %s", intent_id, status)
        return

    raw_payload       = intent.get("payload") or {}
    effective_payload = raw_payload.get("parent_payload") or raw_payload

    result = _process_webhook(effective_payload)
    log.info("processed webhook for %s: fingerprint=%s", intent_id, result["fingerprint"])

    try:
        client.resume_intent(
            intent_id,
            {"action": "complete", **result},
            owner_agent=AXME_AGENT_ADDRESS,
        )
        log.info("resumed intent %s", intent_id)
    except Exception as exc:
        log.error("resume_intent(%s) failed: %s", intent_id, exc)


def main() -> None:
    client = AxmeClient(AxmeClientConfig(api_key=AXME_API_KEY, base_url=AXME_BASE_URL))
    log.info("webhook-receiver starting  address=%s  binding=stream", AXME_AGENT_ADDRESS)

    try:
        for delivery in client.listen(AXME_AGENT_ADDRESS, timeout_seconds=None):
            handle_intent(client, delivery)
    except KeyboardInterrupt:
        log.info("shutting down")


if __name__ == "__main__":
    main()

File: examples/durability/retry-failure/typescript/agent.ts
/**
 * Webhook Receiver Agent — durability/retry-failure example (TypeScript).
 *
 * Processes webhook events idempotently. Run WITHOUT the agent to observe
 * AXME's retry-then-fail behavior.
 *
 * Run:
 *   export AXME_API_KEY=<your-api-key>
 *   export AXME_AGENT_ADDRESS=webhook-receiver-agent
 *   npx tsx examples/durability/retry-failure/agent.ts
 */
import { createHash } from "node:crypto";
import { AxmeClient } from "@axme/axme";

const AXME_BASE_URL = process.env.AXME_BASE_URL ?? "https://api.cloud.axme.ai";
const AXME_API_KEY = process.env.AXME_API_KEY ?? "";
const AXME_AGENT_ADDRESS = process.env.AXME_AGENT_ADDRESS ?? "webhook-receiver-agent";

if (!AXME_API_KEY) { console.error("AXME_API_KEY is required"); process.exit(1); }

function processWebhook(payload: Record<string, unknown>): Record<string, unknown> {
  const eventType = String(payload.event_type ?? "unknown");
  const source = String(payload.source ?? "");
  const testId = String(payload.test_id ?? "");
  const fingerprint = createHash("sha256").update(`${eventType}:${source}:${testId}`).digest("hex").slice(0, 16);
  return { event_type: eventType, source, fingerprint, processed: true, note: `webhook event '${eventType}' processed from '${source}'` };
}

const client = new AxmeClient({ apiKey: AXME_API_KEY, baseUrl: AXME_BASE_URL });
console.log(`webhook-receiver starting  address=${AXME_AGENT_ADDRESS}  binding=stream`);

try {
  for await (const delivery of client.listen(AXME_AGENT_ADDRESS)) {
    const intentId = String(delivery.intent_id ?? "");
    if (!intentId) continue;
    console.log(`received intent ${intentId}`);

    let intent: Record<string, unknown>;
    try {
      const resp = await client.getIntent(intentId);
      intent = (resp.intent ?? resp) as Record<string, unknown>;
    } catch (e) { console.error(`get_intent(${intentId}) failed:`, e); continue; }

    const status = String(intent.lifecycle_status ?? intent.status ?? "").toUpperCase();
    if (!["CREATED", "DELIVERED", "ACKNOWLEDGED", "IN_PROGRESS", "WAITING"].includes(status)) continue;

    const rawPayload = (intent.payload ?? {}) as Record<string, unknown>;
    const effectivePayload = (rawPayload.parent_payload ?? rawPayload) as Record<string, unknown>;
    const result = processWebhook(effectivePayload);

    console.log(`processed webhook for ${intentId}: fingerprint=${result.fingerprint}`);

    try {
      await client.resumeIntent(intentId, { action: "complete", ...result }, { ownerAgent: AXME_AGENT_ADDRESS });
      console.log(`resumed intent ${intentId}`);
    } catch (e) { console.error(`resume_intent(${intentId}) failed:`, e); }
  }
} catch (e) {
  if ((e as Error).name !== "AbortError") throw e;
}
